lib/src/Uncertainty/Process/DiscreteMarkovChain.cxx

```diff
@@ -0,0 +1,243 @@
+ *  it under the terms of the GNU Lesser General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This library is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "openturns/DiscreteMarkovChain.hxx"
+#include "openturns/PersistentObjectFactory.hxx"
+#include "openturns/UserDefined.hxx"
+#include "openturns/ResourceMap.hxx"
+#include "openturns/Point.hxx"
+#include "openturns/SpecFunc.hxx"
+
+
+BEGIN_NAMESPACE_OPENTURNS
+
+CLASSNAMEINIT(DiscreteMarkovChain)
+
+static const Factory<DiscreteMarkovChain> Factory_DiscreteMarkovChain;
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain()
+  : ProcessImplementation()
+  , origin_(0)
+  , transitionMatrix_(SquareMatrix(1, 1))
+  , currentState_(0)
+{
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  // Set the dimension of the process
+  setOutputDimension(1);
+}
+
+/* Standard constructor */
+DiscreteMarkovChain::DiscreteMarkovChain(const UnsignedInteger origin,
+                       const SquareMatrix & transitionMatrix,
+                       const RegularGrid & timeGrid)
+  : ProcessImplementation()
+  , origin_(origin)
+  , transitionMatrix_(transitionMatrix)
+  , currentState_(origin)
+{
+  setOutputDimension(transitionMatrix.getNbRows());
+  setTimeGrid(timeGrid);
+}
+
+/* Virtual constructor */
+DiscreteMarkovChain * DiscreteMarkovChain::clone() const
+{
+  return new DiscreteMarkovChain(*this);
+}
+
+/* String converter */
+String DiscreteMarkovChain::__repr__() const
+{
+  OSS oss;
+  oss << "class=" << DiscreteMarkovChain::GetClassName()
+      << ", origin=" << origin_
+      << ", transition matrix=" << transitionMatrix_;
+  return oss;
+}
+
+/* String converter */
+String DiscreteMarkovChain::__str__() const
+{
+  OSS oss;
+  oss << getClassName() << "(origin=" << origin_ << ", transition=" << PersistentCollection<Scalar>(*transitionMatrix_.getImplementation()).__str__() << ")";
+  return oss;
+}
+
+/* Realization accessor */
+Field DiscreteMarkovChain::getRealization() const
+{
+  const UnsignedInteger size = mesh_.getVerticesNumber();
+  const UnsignedInteger dimension = transitionMatrix_.getNbRows();
+
+  
+  /* Definition of the states of the system */
+  Sample states(dimension, 1);
```

Comment:
The lines from 103 to 126 should be replaced by:
```
  Indices states(dimension);
  states.fill();

  /* Loop on mesh points */
  Sample data(size, 1);
  data(0, 0) = origin_;
  for (UnsignedInteger i = 1; i < size; ++i)
  {
    /* Definition of the distribution */
    Point weights(dimension, 0.0);
    for (UnsignedInteger col = 0; col < dimension; ++col)
    {
      weights[col] = transitionMatrix_(currentState_, col);
    }
    /* Get a realization of the distribution */
    data(i, 0) = DistFunc::rDiscrete(weights);
    currentState_ = data(i, 0);
  }
```
The use of DistFunc::rDiscrete() instead of creating many UserDefined and the use of native types (Indices vs Sample, avoid conversions from UnsignedInteger to Scalar then to UnsignedInteger) lead to a speedup of 6.7 on this benchmark:
```
transitionMatrix = ot.SquareMatrix([[0.0, 0.5, 0.5], [0.7, 0.0, 0.3], [0.8, 0.0, 0.2]])
origin = 1
process = ot.DiscreteMarkovChain(origin, transitionMatrix)
process.setTimeGrid(ot.RegularGrid(0, 1, 200))

from time import time
t0 = time()
sample = process.getSample(1000)
t1 = time()
print("t=", t1 - t0, "s")
```
The initial time was 0.934s, it drops to 0.138s with the proposed change.